filecoin/index.js

```diff
@@ -0,0 +1,80 @@
+import { GetObjectCommand, } from '@aws-sdk/client-s3'
+
+import * as Hasher from 'fr32-sha2-256-trunc254-padded-binary-tree-multihash'
+import * as Digest from 'multiformats/hashes/digest'
+import { Piece } from '@web3-storage/data-segment'
+import { CID } from 'multiformats/cid'
+
+import { GetCarFailed, ComputePieceFailed } from './errors.js'
+
+/**
+ * @typedef {object} EventRecord
+ * @property {string} bucketRegion
+ * @property {string} bucketName
+ * @property {string} key
+ *
+ * @typedef {import('@aws-sdk/client-s3').S3Client} S3Client
+ * @typedef {import('@aws-sdk/client-dynamodb').DynamoDBClient} DynamoDBClient
+ * 
+ * @param {DynamoDBClient} props.dynamoClient
+ * @param {string} props.pieceTableName
+ */
+
+/**
+ * Create CAR side index and write it to Satnav bucket if non existing.
+ *
+ * @param {object} props
+ * @param {EventRecord} props.record
+ * @param {S3Client} props.s3Client
+ * @param {import('./types').PieceTable} props.pieceTable
+ */
+export async function computePieceCid({
+  record,
+  s3Client,
+  pieceTable
+}) {
+  const key = record.key
+  // CIDs in carpark are in format `${carCid}/${carCid}.car`
+  const cidString = key.split('/')[0]
+
+  const getCmd = new GetObjectCommand({
+    Bucket: record.bucketName,
+    Key: key,
+  })
+  const res = await s3Client.send(getCmd)
+  if (!res.Body) {
+    return {
+      error: new GetCarFailed(`failed to get CAR file with key ${key} in bucket ${record.bucketName}`)
+    }
+  }
```

Comment:
To catch the thrown 404 (and others) and return a nicer error:

```suggestion
  let res
  try {
    res = await s3Client.send(getCmd)
    if (!res.Body) throw new Error('missing body')
  } catch (err) {
    return {
      error: new GetCarFailed(`failed to get CAR: ${record.bucketName}/${key}`, { cause: err })
    }
  }
```